flink-connectors/flink-connector-aws-kinesis-data-streams/src/main/java/org/apache/flink/connector/kinesis/sink/KinesisDataStreamsSinkWriter.java

```diff
@@ -0,0 +1,204 @@
+            long flushOnBufferSizeInBytes,
+            long maxTimeInBufferMS,
+            boolean failOnError,
+            String streamName,
+            Properties kinesisClientProperties) {
+        super(
+                elementConverter,
+                context,
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                flushOnBufferSizeInBytes,
+                maxTimeInBufferMS);
+        this.failOnError = failOnError;
+        this.streamName = streamName;
+        this.metrics = context.metricGroup();
+        this.numRecordsOutErrorsCounter = metrics.getNumRecordsOutErrorsCounter();
+        this.client = buildClient(kinesisClientProperties);
+    }
+
+    private KinesisAsyncClient buildClient(Properties kinesisClientProperties) {
+        final ClientConfiguration clientConfiguration =
+                new ClientConfigurationFactory().getConfig();
+        clientConfiguration.setUseTcpKeepAlive(true);
+
+        final SdkAsyncHttpClient httpClient =
+                AwsV2Util.createHttpClient(
+                        clientConfiguration,
+                        NettyNioAsyncHttpClient.builder(),
+                        kinesisClientProperties);
+
+        return AwsV2Util.createKinesisAsyncClient(
+                kinesisClientProperties, clientConfiguration, httpClient);
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<PutRecordsRequestEntry> requestEntries,
+            Consumer<Collection<PutRecordsRequestEntry>> requestResult) {
+
+        PutRecordsRequest batchRequest =
+                PutRecordsRequest.builder().records(requestEntries).streamName(streamName).build();
+
+        LOG.trace("Request to submit {} entries to KDS using KDS Sink.", requestEntries.size());
+
+        CompletableFuture<PutRecordsResponse> future = client.putRecords(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
+                    if (err != null) {
+                        handleFullyFailedRequest(
+                                err, requestEntries, requestResult, getFatalExceptionCons());
+                    } else if (response.failedRecordCount() > 0) {
+                        handlePartiallyFailedRequest(
+                                response, requestEntries, requestResult, getFatalExceptionCons());
+                    } else {
+                        requestResult.accept(Collections.emptyList());
+                    }
+                });
+    }
+
+    @Override
+    protected long getSizeInBytes(PutRecordsRequestEntry requestEntry) {
+        return requestEntry.data().asByteArrayUnsafe().length;
+    }
+
+    private void handleFullyFailedRequest(
+            Throwable err,
+            List<PutRecordsRequestEntry> requestEntries,
+            Consumer<Collection<PutRecordsRequestEntry>> requestResult,
+            Consumer<Exception> exceptionConsumer) {
```

Comment:
Nit: the parameter is not necessary because you can access the instance method getFatalExceptionCons directly.